File: apps/ios/Exponential/Data/API/AgentPlanApi.swift
import Foundation

// Mirrors apps/web/src/lib/trpc/agent-plan.ts. The daemon submits plans via
// the same router; the iOS surface here only exposes the human-side
// approve / request-changes / retry actions.
private struct IssueIdInput: Encodable {
    let issueId: String
}

private struct EmptyResult: Decodable {}

final class AgentPlanApi: Sendable {
    private let trpc: TrpcClient

    init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    func approvePlan(accountId: String, issueId: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "agentPlan.approvePlan",
            input: IssueIdInput(issueId: issueId)
        )
    }

    func requestChanges(accountId: String, issueId: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "agentPlan.requestChanges",
            input: IssueIdInput(issueId: issueId)
        )
    }

    func retry(accountId: String, issueId: String) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "agentPlan.retry",
            input: IssueIdInput(issueId: issueId)
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/api/AgentPlanApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Mirrors apps/web/src/lib/trpc/agent-plan.ts. The agent submits plans via the
// same router; this Android surface exposes the human-side actions (approve /
// request-changes / retry / answer a question). There is no `getState` read
// anymore: every field the Plan Panel needs (plan text, question, revision,
// approval, lastError) is synced locally via the `agent_runs` Electric shape.
@Serializable
private data class IssueIdInput(@SerialName("issueId") val issueId: String)

@Serializable
private data class AnswerInput(
    @SerialName("issueId") val issueId: String,
    @SerialName("answer") val answer: String,
)

@Singleton
class AgentPlanApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun approvePlan(accountId: String, issueId: String) {
        trpc.mutationUnit(
            accountId,
            path = "agentPlan.approvePlan",
            input = IssueIdInput(issueId),
            inputSerializer = IssueIdInput.serializer(),
        )
    }

    suspend fun requestChanges(accountId: String, issueId: String) {
        trpc.mutationUnit(
            accountId,
            path = "agentPlan.requestChanges",
            input = IssueIdInput(issueId),
            inputSerializer = IssueIdInput.serializer(),
        )
    }

    suspend fun retry(accountId: String, issueId: String) {
        trpc.mutationUnit(
            accountId,
            path = "agentPlan.retry",
            input = IssueIdInput(issueId),
            inputSerializer = IssueIdInput.serializer(),
        )
    }

    /** Record the human's answer to the agent's open question. The server clears
     *  the question, flips the issue to `drafting`, and records an agent_answer
     *  event. */
    suspend fun answerQuestion(accountId: String, issueId: String, answer: String) {
        trpc.mutationUnit(
            accountId,
            path = "agentPlan.answerQuestion",
            input = AnswerInput(issueId, answer),
            inputSerializer = AnswerInput.serializer(),
        )
    }
}
